README.md
一款封装系统 Table 的自用的Flutter 表格组件

## 安装

```shell
flutter pub add flutter_app_table
```

## 使用

导入库

```dart
import 'package:flutter_app_table/flutter_app_table.dart';
```

### AppTable

```dart
Widget build(BuildContext context) {
    return AppTable(
        columnCount: 3, // 设置列(必须)
    );
}
```

#### AppTable 具体属性说明

| 字段                 | 类型                         | 说明                 |
| -------------------- | ---------------------------- | -------------------- |
| columnCount          | int                          | 设置表格列数【必须】 |
| rowCount             | int                          | 设置表格行数         |
| defaultColumnWidth   | double                       | 设置默认每一列的宽度 |
| rowBuilder           | AppTableRowBuilder           | 构建表格行内容       |
| headerBuilder        | AppTableColumnBuilder        | 构建表头的内容       |
| columnWidthBuilder   | AppTableColumnWidthBuilder   | 自定每一列的宽度     |
| headerDecoration     | Decoration                   | 表头的外观           |
| rowDecorationBuilder | AppTableRowDecorationBuilder | 每一行的外观设置     |
| border               | TableBorder                  | 表格边框             |
| controller           | AppTableController           | 表格数据源控制器     |

#### AppTableCell 属性说明

| 字段            | 类型              | 说明                       |
| --------------- | ----------------- | -------------------------- |
| builder         | WidgetBuilder     | 绘制Cell组件               |
| height          | double            | 单元格的高度默认40         |
| backgroundColor | Color             | 单元格背景颜色默认白色     |
| selectRowColor  | Color             | 单元格选中时的背景颜色     |
| alignment       | AlignmentGeometry | 布局方式 默认居中          |
| padding         | EdgeInsets        | 设置单元格的内边距默认为 0 |

##### 工厂方法

空的

```dart
AppTableCell.empty()
```

文本展示

```dart
AppTableCell.text({
  required String text,
  TextStyle textStyle = const TextStyle(
    fontSize: 14,
    color: Color(0xFF333333),
  ),
  double height = 40,
  Color backgroundColor = const Color(0xFFFFFFFF),
  Color? selectRowColor,
  AlignmentGeometry alignment = Alignment.center,
  EdgeInsets padding = const EdgeInsets.all(0),
  Key? key,
})
```

可编辑

```dart
AppTableCell.textEdit({
  required String text,
  TextStyle textStyle = const TextStyle(
    fontSize: 14,
    color: Color(0xFF333333),
  ),
  double height = 40,
  Color backgroundColor = const Color(0xFFFFFFFF),
  TextEditingController? controller,
  Color? selectRowColor,
  AlignmentGeometry alignment = Alignment.center,
  EdgeInsets padding = const EdgeInsets.all(0),
  List<TextInputFormatter> inputFormatters = const [],
  TextInputType? keyboardType,
  Key? key,
  ValueChanged<String?>? onSubmitted,
})
```



CHANGELOG.md
## 0.1.0

* 具体参考 [README](./README.md)


lib/app_table/app_table.dart
import 'package:flutter/material.dart';
import 'package:flutter_app_table/app_table/app_table_cell.dart';
import 'package:flutter_app_table/app_table/app_table_controller.dart';
import 'package:provider/provider.dart';

typedef AppTableColumnBuilder = AppTableCell Function(int column);
typedef AppTableRowBuilder = AppTableCell Function(int row, int column);
typedef AppTableColumnWidthBuilder = TableColumnWidth Function(int column);
typedef AppTableRowDecorationBuilder = Decoration? Function(int column);

/// 基于 Table 的表格封装
class AppTable extends StatefulWidget {
  /// 列数
  final int columnCount;

  /// 行数
  final int rowCount;

  /// 默认列宽度 默认为100
  final double defaultColumnWidth;

  /// 构建表内容
  final AppTableRowBuilder? rowBuilder;

  /// 构建表头
  final AppTableColumnBuilder? headerBuilder;

  /// 表头宽度
  final AppTableColumnWidthBuilder? columnWidthBuilder;

  final Decoration? headerDecoration;
  final AppTableRowDecorationBuilder? rowDecorationBuilder;

  /// 表格边框
  final TableBorder? border;

  /// 表格控制器
  final AppTableController? controller;

  const AppTable({
    Key? key,
    required this.columnCount,
    this.rowCount = 0,
    this.defaultColumnWidth = 100,
    this.headerBuilder,
    this.rowBuilder,
    this.columnWidthBuilder,
    this.border,
    this.controller,
    this.headerDecoration,
    this.rowDecorationBuilder,
  }) : super(key: key);

  @override
  State<AppTable> createState() => _AppTableState();
}

class _AppTableState extends State<AppTable> {
  late AppTableController _controller;

  @override
  void initState() {
    super.initState();
    _controller = widget.controller ?? AppTableController();
  }

  @override
  void didUpdateWidget(covariant AppTable oldWidget) {
    super.didUpdateWidget(oldWidget);
    final tableController = widget.controller ?? _controller;
    if (tableController != _controller) {
      _controller = tableController;
    }
  }

  @override
  Widget build(BuildContext context) {
    /// 修复 因为设置 columnCount = 0 报错
    if (widget.columnCount == 0) return Container();
    return Material(
      color: Colors.transparent,
      child: Table(
        border: widget.border ?? _normalBorder,
        columnWidths: _columnWidths,
        children: _cellContents(context),
      ),
    );
  }

  TableBorder get _normalBorder {
    return TableBorder.all(
      color: const Color(0xFFE5E5E5),
      width: 1,
    );
  }

  Map<int, TableColumnWidth> get _columnWidths {
    final Map<int, TableColumnWidth> columnWidths = {};
    for (var i = 0; i < widget.columnCount; i++) {
      columnWidths[i] = _getColumnWidth(i);
    }
    return columnWidths;
  }

  /// 根据 column 获取宽度
  TableColumnWidth _getColumnWidth(int column) {
    final columnWidthBuilder = widget.columnWidthBuilder;
    if (columnWidthBuilder != null) {
      return columnWidthBuilder(column);
    }
    return FixedColumnWidth(widget.defaultColumnWidth);
  }

  /// 获取布局的Cell 内容
  List<TableRow> _cellContents(BuildContext context) {
    List<TableRow> rows = [];

    /// 构建 Header
    final headerBuilder = widget.headerBuilder;
    if (headerBuilder != null) {
      rows.add(TableRow(
        decoration: widget.headerDecoration,
        children: List.generate(widget.columnCount, (column) {
          return _cellContent(
            notifier: _controller.columnNotifier(column),
            context: context,
            cellBuilder: () => headerBuilder(column),
          );
        }),
      ));
    }

    final rowBuilder = widget.rowBuilder;
    if (rowBuilder != null) {
      rows.addAll(List.generate(widget.rowCount, (index) {
        return TableRow(
          decoration: widget.rowDecorationBuilder?.call(index),
          children: List.generate(widget.columnCount, (column) {
            return _cellContent(
              notifier: _controller.rowNotifier(index, column),
              context: context,
              cellBuilder: () => rowBuilder(index, column),
            );
          }),
        );
      }));
    }

    return rows;
  }

  Widget _cellContent({
    required ValueNotifier<int> notifier,
    required BuildContext context,
    required AppTableCell Function() cellBuilder,
  }) {
    return ChangeNotifierProvider.value(
      value: notifier,
      builder: (context, child) {
        return Consumer<ValueNotifier<int>>(
          builder: (context, value, child) {
            final cell = cellBuilder();
            return Container(
              height: cell.height,
              color: cell.backgroundColor,
              width: double.infinity,
              padding: cell.padding,
              child: Align(
                alignment: cell.alignment,
                child: cell.builder(context),
              ),
            );
          },
        );
      },
    );
  }
}


lib/app_table/app_table_edit_cell.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// 可以进行编辑的 Cell
class AppTableEditCell extends StatefulWidget {
  /// 内容文本
  final String text;

  /// text edit controller
  final TextEditingController? controller;

  /// 输入框输入文本样式
  final List<TextInputFormatter> inputFormatters;

  final TextInputType? keyboardType;

  /// 显示文本的样式
  final TextStyle textStyle;

  final ValueChanged<String?>? onSubmitted;

  const AppTableEditCell({
    Key? key,
    required this.text,
    this.controller,
    this.textStyle = const TextStyle(color: Color(0xFF333333), fontSize: 14),
    this.inputFormatters = const [],
    this.keyboardType,
    this.onSubmitted,
  }) : super(key: key);

  @override
  State<AppTableEditCell> createState() => _AppTableEditCellState();
}

class _AppTableEditCellState extends State<AppTableEditCell> {
  late TextEditingController _controller;
  bool isEditing = false;
  final FocusNode _focusNode = FocusNode();

  @override
  void initState() {
    super.initState();
    _controller = widget.controller ??
        TextEditingController.fromValue(TextEditingValue(text: widget.text));

    _focusNode.addListener(() {
      if (!_focusNode.hasFocus) {
        isEditing = false;
        setState(() {});
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: () {
        setState(() {
          isEditing = !isEditing;
        });
        _focusNode.requestFocus();
      },
      child: SizedBox(
        width: double.infinity,
        height: double.infinity,
        child: isEditing ? _buildEditCell(context) : _buildTextCell(context),
      ),
    );
  }

  /// 显示输入框
  Widget _buildEditCell(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
        border: Border.all(
          color: const Color(0xFF00AA91),
          width: isEditing ? 1 : 0,
        ),
      ),
      child: TextField(
        controller: _controller,
        textAlign: TextAlign.center,
        readOnly: !isEditing,
        decoration: InputDecoration(
          border: InputBorder.none,
          hintText: widget.text,
          hintStyle: widget.textStyle,
        ),
        style: widget.textStyle,
        focusNode: _focusNode,
        inputFormatters: widget.inputFormatters,
        keyboardType: widget.keyboardType,
        onSubmitted: (value) {
          isEditing = !isEditing;
          setState(() {});
          widget.onSubmitted?.call(value);
        },
      ),
    );
  }

  Widget _buildTextCell(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Text(
          widget.text,
          style: const TextStyle(color: Color(0xFF333333), fontSize: 14),
        ),
        const SizedBox(width: 10),
        const Icon(
          Icons.mode_edit_outlined,
          size: 12,
          color: Color(0xFF00AA91),
        ),
      ],
    );
  }
}


lib/app_table/app_table_controller.dart
import 'package:flutter/material.dart';
import 'package:flutter_app_table/app_table/app_table_index_path.dart';

/// AppTable 控制器
class AppTableController {
  /// 表格单元格数据
  final Map<String, ValueNotifier<int>> _cellMap = {};

  AppTableController();

  /// 获取 header 对应 Column 数据
  /// [column] 列索引
  /// [row] 行索引
  ValueNotifier<int> columnNotifier(int column) {
    final key = rowCellKey(-1, column);
    final notifier = ValueNotifier(0);
    _cellMap[key] = notifier;
    return notifier;
  }

  /// 获取 row 对应 Row 数据
  /// [row] 行索引
  /// [column] 列索引
  ValueNotifier<int> rowNotifier(int row, int column) {
    final key = rowCellKey(row, column);
    final notifier = ValueNotifier(0);
    _cellMap[key] = notifier;
    return notifier;
  }

  String rowCellKey(int row, int column) {
    return 'row-$row-column-$column';
  }

  /// 刷新单个单元格
  /// [indexPaths] 单元格索引
  void reloadRow(List<AppTableIndexPath> indexPaths) {
    for (var indexPath in indexPaths) {
      final key = rowCellKey(indexPath.row, indexPath.column);
      final notifier = _cellMap[key];
      if (notifier != null) {
        notifier.value = notifier.value + 1;
      }
    }
  }

  /// 刷新 header
  /// [indexPaths] 单元格索引
  void reloadHeader(List<int> columns) {
    for (var column in columns) {
      final key = rowCellKey(-1, column);
      final notifier = _cellMap[key];
      if (notifier != null) {
        notifier.value = notifier.value + 1;
      }
    }
  }
}


lib/app_table/app_table_cell.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'app_table_edit_cell.dart';

class AppTableCell {
  /// 绘制Cell组件
  final WidgetBuilder builder;

  /// 单元格的宽度
  final double height;

  /// 单元格背景颜色
  final Color backgroundColor;

  /// 单元格选中时的背景颜色
  final Color? selectRowColor;

  /// 布局方式 默认居中
  final AlignmentGeometry alignment;

  /// 设置单元格的内边距
  final EdgeInsets padding;

  /// 默认初始化
  /// [builder] 自定义绘制Cell组件
  /// [height] 单元格的高度
  /// [backgroundColor] 单元格的背景颜色
  /// [selectRowColor] 单元格选中时的背景颜色
  AppTableCell({
    required this.builder,
    this.height = 40,
    this.backgroundColor = const Color(0xFFFFFFFF),
    this.selectRowColor,
    this.alignment = Alignment.center,
    this.padding = const EdgeInsets.all(0),
  });

  /// 创建一个空的Cell
  factory AppTableCell.empty() => AppTableCell(builder: (_) => Container());

  /// 创建一个展示文本的 Cell
  /// [text] 文本内容
  /// [textStyle] 文本样式
  /// [height] 文本对齐方式
  /// [backgroundColor] 单元格的背景颜色
  /// [selectRowColor] 单元格选中时的背景颜色
  factory AppTableCell.text({
    required String text,
    TextStyle textStyle = const TextStyle(
      fontSize: 14,
      color: Color(0xFF333333),
    ),
    double height = 40,
    Color backgroundColor = const Color(0xFFFFFFFF),
    Color? selectRowColor,
    AlignmentGeometry alignment = Alignment.center,
    EdgeInsets padding = const EdgeInsets.all(0),
    Key? key,
  }) =>
      AppTableCell(
        builder: (context) => Text(text, style: textStyle, key: key),
        height: height,
        backgroundColor: backgroundColor,
        selectRowColor: selectRowColor,
        alignment: alignment,
        padding: padding,
      );

  /// 创建一个可以编辑的Cell
  /// [text] 文本内容
  /// [textStyle] 文本样式
  /// [height] 文本对齐方式
  /// [backgroundColor] 单元格的背景颜色
  /// [controller] 文本控制器
  /// [selectRowColor] 单元格选中时的背景颜色
  factory AppTableCell.textEdit({
    required String text,
    TextStyle textStyle = const TextStyle(
      fontSize: 14,
      color: Color(0xFF333333),
    ),
    double height = 40,
    Color backgroundColor = const Color(0xFFFFFFFF),
    TextEditingController? controller,
    Color? selectRowColor,
    AlignmentGeometry alignment = Alignment.center,
    EdgeInsets padding = const EdgeInsets.all(0),
    List<TextInputFormatter> inputFormatters = const [],
    TextInputType? keyboardType,
    Key? key,
    ValueChanged<String?>? onSubmitted,
  }) =>
      AppTableCell(
        builder: (context) => AppTableEditCell(
          text: text,
          textStyle: textStyle,
          controller: controller,
          inputFormatters: inputFormatters,
          keyboardType: keyboardType,
          key: key,
          onSubmitted: onSubmitted,
        ),
        height: height,
        backgroundColor: backgroundColor,
        selectRowColor: selectRowColor,
        alignment: alignment,
        padding: padding,
      );

  bool _isSelect = false;

  bool get isSelected => _isSelect;

  set isSelected(bool value) {
    if (_isSelect == value) return;
    _isSelect = value;
  }

  Color get selectBackgroundColor => selectRowColor ?? backgroundColor;
}
